File: integration_tests/java_client_tests/client-app/src/test/java/dev/graphlink/test/NullabilityTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.types.FindUserResponse;
import dev.graphlink.test.generated.types.GetPostResponse;
import dev.graphlink.test.generated.types.GetUserResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class NullabilityTest {

    private MockAdapter adapter;
    private GraphLinkClient client;

    @BeforeEach
    void setUp() {
        adapter = new MockAdapter();
        client = newClient(adapter, new MockWebSocketAdapter());
    }

    // ── Non-nullable query return type ────────────────────────────────────────

    @Nested
    class NonNullableQueryReturn {

        @BeforeEach
        void register() {
            adapter.registerData("getUser", Map.of("getUser", kUserAliceJson));
        }

        @Test
        void getUser_returnsNonNullUser() {
            GetUserResponse res = client.queries.getUser("user-1");
            assertNotNull(res.getGetUser());
        }

        @Test
        void getUser_hasCorrectId() {
            assertEquals("user-1", client.queries.getUser("user-1").getGetUser().getId());
        }
    }

    // ── Nullable query return type — null ─────────────────────────────────────

    @Nested
    class NullableQueryReturn_Null {

        @BeforeEach
        void register() {
            adapter.registerData("findUser", map("findUser", null));
        }

        @Test
        void findUser_returnsNull_whenServerReturnsNull() {
            FindUserResponse res = client.queries.findUser("user-99");
            assertNull(res.getFindUser());
        }
    }

    // ── Nullable query return type — present ──────────────────────────────────

    @Nested
    class NullableQueryReturn_Present {

        @BeforeEach
        void register() {
            adapter.registerData("findUser", Map.of("findUser", kUserAliceJson));
        }

        @Test
        void findUser_returnsUser_whenServerReturnsData() {
            FindUserResponse res = client.queries.findUser("user-1");
            assertNotNull(res.getFindUser());
            assertEquals("user-1", res.getFindUser().getId());
        }
    }

    // ── Nullable nested object — null ─────────────────────────────────────────

    @Nested
    class NullableNestedObject_Null {

        @BeforeEach
        void register() {
            adapter.registerData("getUser", Map.of("getUser", kUserAliceJson));
        }

        @Test
        void billingAddress_isNull_whenAbsent() {
            assertNull(client.queries.getUser("user-1").getGetUser().getBillingAddress());
        }
    }

    // ── Nullable nested object — present ──────────────────────────────────────

    @Nested
    class NullableNestedObject_Present {

        @BeforeEach
        void register() {
            adapter.registerData("getUser", Map.of("getUser", kUserBobJson));
        }

        @Test
        void billingAddress_isNotNull_whenPresent() {
            assertNotNull(client.queries.getUser("user-2").getGetUser().getBillingAddress());
        }

        @Test
        void billingAddress_hasCorrectCity() {
            assertEquals("Capital City",
                    client.queries.getUser("user-2").getGetUser().getBillingAddress().getCity());
        }
    }

    // ── Nullable scalar on nested type ────────────────────────────────────────

    @Nested
    class NullableScalarOnNestedType {

        @Test
        void address_zip_isNull_whenAbsent() {
            adapter.registerData("getUser", Map.of("getUser", kUserBobJson));
            assertNull(client.queries.getUser("user-2").getGetUser().getAddress().getZip());
        }

        @Test
        void address_zip_isPresent_whenSet() {
            adapter.registerData("getUser", Map.of("getUser", kUserAliceJson));
            assertEquals("12345",
                    client.queries.getUser("user-1").getGetUser().getAddress().getZip());
        }
    }

    // ── Nullable list field ───────────────────────────────────────────────────

    @Nested
    class NullableListField {

        @Test
        void scores_isNull_whenAbsent() {
            adapter.registerData("getUser", Map.of("getUser", kUserBobJson));
            assertNull(client.queries.getUser("user-2").getGetUser().getScores());
        }

        @Test
        void scores_hasList_whenPresent() {
            adapter.registerData("getUser", Map.of("getUser", kUserAliceJson));
            assertNotNull(client.queries.getUser("user-1").getGetUser().getScores());
            assertEquals(3, client.queries.getUser("user-1").getGetUser().getScores().size());
        }
    }

    // ── Nullable object on Post — coAuthor ────────────────────────────────────

    @Nested
    class NullableCoAuthor {

        @Test
        void coAuthor_isNull_whenAbsent() {
            adapter.registerData("getPost", Map.of("getPost", kPostJson));
            assertNull(client.queries.getPost("post-1").getGetPost().getCoAuthor());
        }

        @Test
        void coAuthor_isNotNull_whenPresent() {
            adapter.registerData("getPost", Map.of("getPost", kPostWithCoAuthorJson));
            assertNotNull(client.queries.getPost("post-2").getGetPost().getCoAuthor());
            assertEquals("user-2",
                    client.queries.getPost("post-2").getGetPost().getCoAuthor().getId());
        }
    }
}

File: integration_tests/kotlin_client_tests/client-app/src/test/kotlin/dev/graphlink/test/NullabilityTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.kPostJson
import dev.graphlink.test.Fixtures.kPostWithCoAuthorJson
import dev.graphlink.test.Fixtures.kUserAliceJson
import dev.graphlink.test.Fixtures.kUserBobJson
import dev.graphlink.test.Fixtures.newClient
import dev.graphlink.test.n

class NullabilityTest {

    private lateinit var adapter: MockAdapter
    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        adapter = MockAdapter()
        client = newClient(adapter, MockWebSocketAdapter())
    }

    // ── Non-nullable query return type ────────────────────────────────────────

    @Nested
    inner class NonNullableQueryReturn {

        @BeforeTest
        fun register() {
            adapter.registerData("getUser", mapOf("getUser" to kUserAliceJson))
        }

        @Test
        fun getUser_returnsNonNullUser() = runTest {
            assertNotNull(client.queries.getUser("user-1").getUser)
        }

        @Test
        fun getUser_hasCorrectId() = runTest {
            assertEquals("user-1", client.queries.getUser("user-1").getUser.id)
        }
    }

    // ── Nullable query return type — null ─────────────────────────────────────

    @Nested
    inner class NullableQueryReturn_Null {

        @BeforeTest
        fun register() {
            adapter.registerData("findUser", n("findUser" to null))
        }

        @Test
        fun findUser_returnsNull_whenServerReturnsNull() = runTest {
            assertNull(client.queries.findUser("user-99").findUser)
        }
    }

    // ── Nullable query return type — present ──────────────────────────────────

    @Nested
    inner class NullableQueryReturn_Present {

        @BeforeTest
        fun register() {
            adapter.registerData("findUser", mapOf("findUser" to kUserAliceJson))
        }

        @Test
        fun findUser_returnsUser_whenServerReturnsData() = runTest {
            val res = client.queries.findUser("user-1")
            assertNotNull(res.findUser)
            assertEquals("user-1", res.findUser!!.id)
        }
    }

    // ── Nullable nested object — null ─────────────────────────────────────────

    @Nested
    inner class NullableNestedObject_Null {

        @BeforeTest
        fun register() {
            adapter.registerData("getUser", mapOf("getUser" to kUserAliceJson))
        }

        @Test
        fun billingAddress_isNull_whenAbsent() = runTest {
            assertNull(client.queries.getUser("user-1").getUser.billingAddress)
        }
    }

    // ── Nullable nested object — present ──────────────────────────────────────

    @Nested
    inner class NullableNestedObject_Present {

        @BeforeTest
        fun register() {
            adapter.registerData("getUser", mapOf("getUser" to kUserBobJson))
        }

        @Test
        fun billingAddress_isNotNull_whenPresent() = runTest {
            assertNotNull(client.queries.getUser("user-2").getUser.billingAddress)
        }

        @Test
        fun billingAddress_hasCorrectCity() = runTest {
            assertEquals("Capital City", client.queries.getUser("user-2").getUser.billingAddress!!.city)
        }
    }

    // ── Nullable scalar on nested type ────────────────────────────────────────

    @Nested
    inner class NullableScalarOnNestedType {

        @Test
        fun address_zip_isNull_whenAbsent() = runTest {
            adapter.registerData("getUser", mapOf("getUser" to kUserBobJson))
            assertNull(client.queries.getUser("user-2").getUser.address.zip)
        }

        @Test
        fun address_zip_isPresent_whenSet() = runTest {
            adapter.registerData("getUser", mapOf("getUser" to kUserAliceJson))
            assertEquals("12345", client.queries.getUser("user-1").getUser.address.zip)
        }
    }

    // ── Nullable list field ───────────────────────────────────────────────────

    @Nested
    inner class NullableListField {

        @Test
        fun scores_isNull_whenAbsent() = runTest {
            adapter.registerData("getUser", mapOf("getUser" to kUserBobJson))
            assertNull(client.queries.getUser("user-2").getUser.scores)
        }

        @Test
        fun scores_hasList_whenPresent() = runTest {
            adapter.registerData("getUser", mapOf("getUser" to kUserAliceJson))
            val scores = client.queries.getUser("user-1").getUser.scores
            assertNotNull(scores)
            assertEquals(3, scores.size)
        }
    }

    // ── Nullable coAuthor on Post ─────────────────────────────────────────────

    @Nested
    inner class NullableCoAuthor {

        @Test
        fun coAuthor_isNull_whenAbsent() = runTest {
            adapter.registerData("getPost", mapOf("getPost" to kPostJson))
            assertNull(client.queries.getPost("post-1").getPost.coAuthor)
        }

        @Test
        fun coAuthor_isNotNull_whenPresent() = runTest {
            adapter.registerData("getPost", mapOf("getPost" to kPostWithCoAuthorJson))
            val coAuthor = client.queries.getPost("post-2").getPost.coAuthor
            assertNotNull(coAuthor)
            assertEquals("user-2", coAuthor.id)
        }
    }
}
